### ISSUE TITLE
Closing file handles early breaks IO::Select

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#10388]
Migrated from [rt.perl.org#75156](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=75156) (status was 'resolved')

Searchable as RT75156$

**Comment by p5pRT** [comment]
### From bre@klaki.net

#### Created by bre@klaki.net



\(re\-posting what I sent to perl5\-porters&#8203;: the IO&#8203;::Select perldoc
should probably recommend perlbug and not perl5\-porters as a venue
for bug\-reports\)\.

I write to you\, because I think I have found and fixed a rather
nasty \(and old\) bug in IO&#8203;::Select&#8203;:

The bug is&#8203;: IO&#8203;::Select currently assumes that an IO&#8203;::Handle will
always have a valid file descriptor\, which is not actually the case\.
If you add a handle and the handle later loses its file descriptor
\(for whatever reason\, this appears to be out of the control of the
programmer\)\, then IO&#8203;::Select will stop working and there is no way
to remove the broken descriptor\, because the add/remove code
assumes fileno\(\) will return something useful\. The workaround in
my code today is to throw away the IO&#8203;::Select object and build a
new one\.

Attached is a sample program which triggers the bug\, and a proposed
modification to IO&#8203;::Select \(NewSelect\.pm\) which I think should fix
the problem\.  In case I fail to attach the files correctly\, they
can also be found here&#8203;:

&nbsp;  http&#8203;://bre\.klaki\.net/programs/Perl\-IO\-Select\-Bug/


Addition&#8203;:

The real\-world impact of this bug\, is to cause most servers written 
using IO&#8203;::Select to go into infinite CPU\-sucking loops when a
watched file\-handle is closed without being removed from IO&#8203;::Select
first\.  My fix does not actually prevent that entirely\, but it does 
give the programmer a mechanism for dealing with the problem and 
makes code slightly more resiliant \- as long as the file handle 
closure is detected \*somewhere\* and removed from IO&#8203;::Select 
\*sometime\* the system will recover\.

<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=high
    module=IO::Select

Site configuration information for perl 5.10.1:

Configured by Debian Project at Fri Apr 23 07:59:14 UTC 2010.

Summary of my perl5 (revision 5 version 10 subversion 1) configuration:
   
  Platform:
    osname=linux, osvers=2.6.24-27-server, archname=i486-linux-gnu-thread-multi
    uname='linux vernadsky 2.6.24-27-server #1 smp fri mar 12 01:45:06 utc 2010 i686 gnulinux '
    config_args='-Dusethreads -Duselargefiles -Dccflags=-DDEBIAN -Dcccdlflags=-fPIC -Darchname=i486-linux-gnu -Dprefix=/usr -Dprivlib=/usr/share/perl/5.10 -Darchlib=/usr/lib/perl/5.10 -Dvendorprefix=/usr -Dvendorlib=/usr/share/perl5 -Dvendorarch=/usr/lib/perl5 -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl/5.10.1 -Dsitearch=/usr/local/lib/perl/5.10.1 -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/share/man/man3 -Dsiteman1dir=/usr/local/man/man1 -Dsiteman3dir=/usr/local/man/man3 -Dman1ext=1 -Dman3ext=3perl -Dpager=/usr/bin/sensible-pager -Uafs -Ud_csh -Ud_ualarm -Uusesfio -Uusenm -DDEBUGGING=-g -Doptimize=-O2 -Duseshrplib -Dlibperl=libperl.so.5.10.1 -Dd_dosuid -des'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=undef, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2 -g',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.4.3', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib /usr/lib64
    libs=-lgdbm -lgdbm_compat -ldb -ldl -lm -lpthread -lc -lcrypt
    perllibs=-ldl -lm -lpthread -lc -lcrypt
    libc=/lib/libc-2.11.1.so, so=so, useshrplib=true, libperl=libperl.so.5.10.1
    gnulibc_version='2.11.1'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -g -L/usr/local/lib -fstack-protector'

Locally applied patches:
    


@INC for perl 5.10.1:
    /etc/perl
    /usr/local/lib/perl/5.10.1
    /usr/local/share/perl/5.10.1
    /usr/lib/perl5
    /usr/share/perl5
    /usr/lib/perl/5.10
    /usr/share/perl/5.10
    /usr/local/lib/site_perl
    .


Environment for perl 5.10.1:
    HOME=/home/bre
    LANG=en_IE.utf8
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/bre/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @jkeenan

On Tue May 18 00&#8203;:35&#8203;:01 2010\, bre@&#8203;klaki\.net wrote&#8203;:
> This is a bug report for perl from bre@&#8203;klaki\.net\,
> generated with the help of perlbug 1\.39 running under perl 5\.10\.1\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
> 
> \(re\-posting what I sent to perl5\-porters&#8203;: the IO&#8203;::Select perldoc
> should probably recommend perlbug and not perl5\-porters as a venue
> for bug\-reports\)\.
> 
> I write to you\, because I think I have found and fixed a rather
> nasty \(and old\) bug in IO&#8203;::Select&#8203;:
> 
> The bug is&#8203;: IO&#8203;::Select currently assumes that an IO&#8203;::Handle will
> always have a valid file descriptor\, which is not actually the case\.
> If you add a handle and the handle later loses its file descriptor
> \(for whatever reason\, this appears to be out of the control of the
> programmer\)\, then IO&#8203;::Select will stop working and there is no way
> to remove the broken descriptor\, because the add/remove code
> assumes fileno\(\) will return something useful\. The workaround in
> my code today is to throw away the IO&#8203;::Select object and build a
> new one\.
> 
> Attached is a sample program which triggers the bug\, and a proposed
> modification to IO&#8203;::Select \(NewSelect\.pm\) which I think should fix
> the problem\.  In case I fail to attach the files correctly\, they
> can also be found here&#8203;:
> 
>   http&#8203;://bre\.klaki\.net/programs/Perl\-IO\-Select\-Bug/
> 
> 
> Addition&#8203;:
> 
> The real\-world impact of this bug\, is to cause most servers written
> using IO&#8203;::Select to go into infinite CPU\-sucking loops when a
> watched file\-handle is closed without being removed from IO&#8203;::Select
> first\.  My fix does not actually prevent that entirely\, but it does
> give the programmer a mechanism for dealing with the problem and
> makes code slightly more resiliant \- as long as the file handle
> closure is detected \*somewhere\* and removed from IO&#8203;::Select
> \*sometime\* the system will recover\.
> 
> 

This older ticket is difficult to evaluate because the attachments the
author intended to attach never made it\.  I went to the web site
provided and downloaded those files\.  With slight modifications\, I tried
them out and am attaching them\.

Nevertheless\, I can't reproduce the poster's complaint about IO&#8203;::Select\,
that is\, the 'remove' method seems to be succeeding\, whereas the
poster's comments suggest that it should be failing\.  See my attachment\,
75156\_select\.t\.

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
### From @jkeenan

\#\!/usr/bin/perl
push @&#8203;INC\, '\.';
use strict;
use IO&#8203;::Select;
use IO&#8203;::Socket&#8203;::INET;
use Test&#8203;::More qw\( no\_plan \);

\# First just create a couple of sockets to play with\.
my $port = 10000\+\($$ % 500\);
my $bad\_server = IO&#8203;::Socket&#8203;::INET\->new\( 
&nbsp;    LocalAddr => '127\.0\.0\.1'\,
&nbsp;    LocalPort => $port\,
&nbsp;    Type => SOCK\_STREAM\,
&nbsp;    Listen => 5\,
\);
my $dumb\_client = IO&#8203;::Socket&#8203;::INET\->new\(
&nbsp;    PeerAddr => '127\.0\.0\.1'\,
&nbsp;    PeerPort => $port\,
&nbsp;    Type => SOCK\_STREAM\,
\);

\# Add them to our select objects\.\.\.
my $oldSel = IO&#8203;::Select\->new\(\);
$oldSel\->add\($bad\_server\);
$oldSel\->add\($dumb\_client\);

\# Make $dumb\_client\->fileno return undef\.
$dumb\_client\->close\(\);
ok\(\! defined\($dumb\_client\->fileno\(\)\)\,
&nbsp;     "fileno now returns undef"\);

\# Trigger and test the bug&#8203;: we cannot remove $dumb\_client \!
my $before = $oldSel\->as\_string\(\);
$oldSel\->remove\($dumb\_client\);
my $after  = $oldSel\->as\_string\(\);
isnt\($before\, $after\,
&nbsp;    "After remove\(\)\, two lines are not the same"\);

**Comment by p5pRT** [comment]
### From @jkeenan

[75156_select_test.pl](https://rt-archive.perl.org/perl5/Ticket/Attachment/1050498/520584/75156_select_test.pl)

**Comment by p5pRT** [comment]
### From @jkeenan

[NewSelect.pm](https://rt-archive.perl.org/perl5/Ticket/Attachment/1050498/520590/NewSelect.pm)

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

On Tue May 18 00&#8203;:35&#8203;:01 2010\, bre@&#8203;klaki\.net wrote&#8203;:
> This is a bug report for perl from bre@&#8203;klaki\.net\,
> generated with the help of perlbug 1\.39 running under perl 5\.10\.1\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
> 
> \(re\-posting what I sent to perl5\-porters&#8203;: the IO&#8203;::Select perldoc
> should probably recommend perlbug and not perl5\-porters as a venue
> for bug\-reports\)\.

I’ve fixed that with commit 9f7d1e4\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @tonycoz

On Tue Jan 03 19&#8203;:06&#8203;:13 2012\, jkeenan wrote&#8203;:
> On Tue May 18 00&#8203;:35&#8203;:01 2010\, bre@&#8203;klaki\.net wrote&#8203;:
> > \(re\-posting what I sent to perl5\-porters&#8203;: the IO&#8203;::Select perldoc
> > should probably recommend perlbug and not perl5\-porters as a venue
> > for bug\-reports\)\.
> > 
> > I write to you\, because I think I have found and fixed a rather
> > nasty \(and old\) bug in IO&#8203;::Select&#8203;:
> > 
> > The bug is&#8203;: IO&#8203;::Select currently assumes that an IO&#8203;::Handle will
> > always have a valid file descriptor\, which is not actually the case\.
> > If you add a handle and the handle later loses its file descriptor
> > \(for whatever reason\, this appears to be out of the control of the
> > programmer\)\, then IO&#8203;::Select will stop working and there is no way
> > to remove the broken descriptor\, because the add/remove code
> > assumes fileno\(\) will return something useful\. The workaround in
> > my code today is to throw away the IO&#8203;::Select object and build a
> > new one\.
> > 
> > Attached is a sample program which triggers the bug\, and a proposed
> > modification to IO&#8203;::Select \(NewSelect\.pm\) which I think should fix
> > the problem\.  In case I fail to attach the files correctly\, they
> > can also be found here&#8203;:
> > 
> >   http&#8203;://bre\.klaki\.net/programs/Perl\-IO\-Select\-Bug/
> > 
> > 
> > Addition&#8203;:
> > 
> > The real\-world impact of this bug\, is to cause most servers written
> > using IO&#8203;::Select to go into infinite CPU\-sucking loops when a
> > watched file\-handle is closed without being removed from IO&#8203;::Select
> > first\.  My fix does not actually prevent that entirely\, but it does
> > give the programmer a mechanism for dealing with the problem and
> > makes code slightly more resiliant \- as long as the file handle
> > closure is detected \*somewhere\* and removed from IO&#8203;::Select
> > \*sometime\* the system will recover\.
> > 
> > 
> 
> This older ticket is difficult to evaluate because the attachments the
> author intended to attach never made it\.  I went to the web site
> provided and downloaded those files\.  With slight modifications\, I tried
> them out and am attaching them\.
> 
> Nevertheless\, I can't reproduce the poster's complaint about IO&#8203;::Select\,
> that is\, the 'remove' method seems to be succeeding\, whereas the
> poster's comments suggest that it should be failing\.  See my attachment\,
> 75156\_select\.t\.

There's a partial fix for this issue in 2e6546ca\, unfortunately that fix
has a couple of bugs&#8203;:

\- it doesn't update the bit vector \($bits\) so the next can\_read\(\) will
attempt to wait on a closed selector

\- it doesn't update $count\, so remove returns failure even though the FH
was \(buggily\) removed

Tony

**Comment by p5pRT** [comment]
### From @tonycoz

On Tue Sep 17 19&#8203;:11&#8203;:58 2013\, tonyc wrote&#8203;:
> There's a partial fix for this issue in 2e6546ca\, unfortunately that fix
> has a couple of bugs&#8203;:
> 
> \- it doesn't update the bit vector \($bits\) so the next can\_read\(\) will
> attempt to wait on a closed selector
> 
> \- it doesn't update $count\, so remove returns failure even though the FH
> was \(buggily\) removed

Tests and fixes attached\, which I'll apply in a couple of days or so\.

Tony

**Comment by p5pRT** [comment]
### From @tonycoz

<details><summary>0002-perl-75156-fix-the-return-value-and-bits-for-removin.patch</summary>

```diff
From b58d6c0df55563f960c7413cd97662e2c38870dc Mon Sep 17 00:00:00 2001
From: Tony Cook <tony@develop-help.com>
Date: Wed, 18 Sep 2013 15:55:12 +1000
Subject: [PATCH 2/2] [perl #75156] fix the return value and bits for removing
 a closed fh

---
 dist/IO/lib/IO/Select.pm |   27 ++++++++++++++++++---------
 dist/IO/t/io_sel.t       |    4 ++--
 2 files changed, 20 insertions(+), 11 deletions(-)

diff --git a/dist/IO/lib/IO/Select.pm b/dist/IO/lib/IO/Select.pm
index 756a299..994f896 100644
--- a/dist/IO/lib/IO/Select.pm
+++ b/dist/IO/lib/IO/Select.pm
@@ -11,7 +11,7 @@ use warnings::register;
 use     vars qw($VERSION @ISA);
 require Exporter;
 
-$VERSION = "1.21";
+$VERSION = "1.22";
 
 @ISA = qw(Exporter); # This is only so we can do version checking
 
@@ -86,15 +86,24 @@ sub _update
      $vec->[$i] = $f;
    } else {      # remove
      if ( ! defined $fn ) { # remove if fileno undef'd
-         defined($_) && $_ == $f and do { $vec->[FD_COUNT]--; $_ = undef; }
-           for @{$vec}[FIRST_FD .. $#$vec];
-         next;
+       $fn = 0;
+       for my $fe (@{$vec}[FIRST_FD .. $#$vec]) {
+         if (defined($fe) && $fe == $f) {
+	   $vec->[FD_COUNT]--;
+	   $fe = undef;
+	   vec($bits, $fn, 1) = 0;
+	   last;
+	 }
+	 ++$fn;
+       }
+     }
+     else {
+       my $i = $fn + FIRST_FD;
+       next unless defined $vec->[$i];
+       $vec->[FD_COUNT]--;
+       vec($bits, $fn, 1) = 0;
+       $vec->[$i] = undef;
      }
-     my $i = $fn + FIRST_FD;
-     next unless defined $vec->[$i];
-     $vec->[FD_COUNT]--;
-     vec($bits, $fn, 1) = 0;
-     $vec->[$i] = undef;
    }
    $count++;
   }
diff --git a/dist/IO/t/io_sel.t b/dist/IO/t/io_sel.t
index bd61b68..34af03a 100644
--- a/dist/IO/t/io_sel.t
+++ b/dist/IO/t/io_sel.t
@@ -143,9 +143,9 @@ print "ok 23\n" ;
     print "ok 24 - added socket\n";
     close $fh;
     print "not " unless $sel->remove($fh) == 1;
-    print "ok 25 - removed closed socket # TODO code doesn't update count\n";
+    print "ok 25 - removed closed socket\n";
     print "not " unless $sel->count == 1;
     print "ok 26 - count() updated\n";
     print "not " unless $sel->bits ne $oldbits;
-    print "ok 27 - bits() updated # TODO code doesn't update bits\n";
+    print "ok 27 - bits() updated\n";
 }
-- 
1.7.10.4


```
</details>

**Comment by p5pRT** [comment]
### From @tonycoz

<details><summary>0001-perl-75156-tests-for-deleting-a-closed-handle-from-I.patch</summary>

```diff
From a78df7b89b5bca007c58dac456521d5b61ff8553 Mon Sep 17 00:00:00 2001
From: Tony Cook <tony@develop-help.com>
Date: Wed, 18 Sep 2013 15:20:19 +1000
Subject: [PATCH 1/2] [perl #75156] tests for deleting a closed handle from
 IO::Select

---
 dist/IO/t/io_sel.t |   22 +++++++++++++++++++++-
 1 file changed, 21 insertions(+), 1 deletion(-)

diff --git a/dist/IO/t/io_sel.t b/dist/IO/t/io_sel.t
index eb4bb90..bd61b68 100644
--- a/dist/IO/t/io_sel.t
+++ b/dist/IO/t/io_sel.t
@@ -3,7 +3,7 @@
 select(STDERR); $| = 1;
 select(STDOUT); $| = 1;
 
-print "1..23\n";
+print "1..27\n";
 
 use IO::Select 1.09;
 
@@ -129,3 +129,23 @@ IO::Select::has_error();
 print "not " unless $w == 1 ;
 $w = 0 ;
 print "ok 23\n" ;
+
+{
+    # perl #75156 - test we can delete a closed handle
+    require IO::Socket::INET;
+    my $fh = IO::Socket::INET->new(
+      Listen => 5,
+    );
+    my $sel = IO::Select->new(\*STDIN);
+    $sel->add($fh);
+    my $oldbits = $sel->bits;
+    print "not " unless $sel->count == 2;
+    print "ok 24 - added socket\n";
+    close $fh;
+    print "not " unless $sel->remove($fh) == 1;
+    print "ok 25 - removed closed socket # TODO code doesn't update count\n";
+    print "not " unless $sel->count == 1;
+    print "ok 26 - count() updated\n";
+    print "not " unless $sel->bits ne $oldbits;
+    print "ok 27 - bits() updated # TODO code doesn't update bits\n";
+}
-- 
1.7.10.4


```
</details>

**Comment by p5pRT** [comment]
### From @tonycoz

On Tue Sep 17 22&#8203;:56&#8203;:31 2013\, tonyc wrote&#8203;:
> On Tue Sep 17 19&#8203;:11&#8203;:58 2013\, tonyc wrote&#8203;:
> > There's a partial fix for this issue in 2e6546ca\, unfortunately that fix
> > has a couple of bugs&#8203;:
> > 
> > \- it doesn't update the bit vector \($bits\) so the next can\_read\(\) will
> > attempt to wait on a closed selector
> > 
> > \- it doesn't update $count\, so remove returns failure even though the FH
> > was \(buggily\) removed
> 
> Tests and fixes attached\, which I'll apply in a couple of days or so\.

Applied as 3bfb0615ce07345b2bdefdd1acbd343f4dfa21d7 and ff4a442c79158c7df99be693a8f8559a09cf6df7\.

Tony

**Comment by p5pRT** [comment]
@tonycoz - Status changed from 'open' to 'resolved'
